haystack/document_stores/opensearch.py

```diff
@@ -699,24 +863,67 @@ def _get_embedding_field_mapping(
             index_type = self.index_type
         if embedding_dim is None:
             embedding_dim = self.embedding_dim
+        if index is None:
+            index = self.index
 
         embeddings_field_mapping = {"type": "knn_vector", "dimension": embedding_dim}
         if knn_engine != "score_script":
-            method: dict = {"space_type": space_type, "name": "hnsw", "engine": knn_engine}
+            method: dict = {"space_type": space_type, "engine": knn_engine}
+
+            ef_construction = (
+                80 if "ef_construction" not in self.knn_parameters else self.knn_parameters["ef_construction"]
+            )
+            ef_search = 20 if "ef_search" not in self.knn_parameters else self.knn_parameters["ef_search"]
+            m = 64 if "m" not in self.knn_parameters else self.knn_parameters["m"]
 
             if index_type == "flat":
+                # As OpenSearch currently does not support all similarity functions (e.g. dot_product) in exact vector
+                # similarity calculations, we don't make use of exact vector similarity when index_type='flat'.
+                # Instead we use the same approximate vector similarity calculations like in 'hnsw', but further
+                # optimized for accuracy.
```

Comment:
It actually supports all similarity functions with exact vector calc, i.e. `score_script` in the meantime. It's kind of ugly what we're doing here with the index_types and knn_engines but it is the smaller evil. Enforcing `score_script` for `flat` index_type would have significant disadvantages:
- we break previous behavior: if you didn't set `index_type` before, you got fast ANN search with high accuracy, now you'd get slow exact vector search
- the `flat` hnsw profile, which is the default setting of OpenSearch, has a very good balance between index size, search latency and accuracy: it wouldn't be there anymore
- you almost never want to use exact knn except for prefiltering (it's just bad as default value)

Anyways, if we want to resolve this dilemma, we should do this in another PR. For now let's just reformulate the comment to something like: we're using HNSW anyways knn_engines nmslib and faiss as they do not support exact knn.